### DISCUSSION TITLE
Roadmap April 2023

### DISCUSSION THREAD

**Discussion by ocots** [control-toolbox/control-toolbox.github.io#34] [General]
See [roadmap March 2023](https://github.com/orgs/control-toolbox/discussions/18)

**Remark.** The direct shooting method will be removed from `OptimalControl` in a future release since it is not enough general.

Main tasks:

- [x] @jbcaillau @j-l-s Abstract formulation working on basic example at least.
- [x] @jbcaillau @joseph-gergaud @PierreMartinon [Box and pure constraints for the direct method](https://github.com/control-toolbox/CTBase.jl/issues/4).
- [ ] @jbcaillau @ocots [Getters of multipliers associated to labeled constraints](https://github.com/control-toolbox/OptimalControl.jl/issues/116).
- [ ] @ocots @joseph-gergaud @PierreMartinon [Functional: Bolza cost](https://github.com/control-toolbox/OptimalControl.jl/issues/82).
- [ ] @joseph-gergaud @PierreMartinon [Direct solve by generic RK scheme](https://github.com/control-toolbox/OptimalControl.jl/issues/117).
- [ ] @joseph-gergaud @PierreMartinon [Sparse usage (Jacobian, Hessian of the constraints)](https://github.com/control-toolbox/OptimalControl.jl/issues/118).
- [ ] @ocots @PierreMartinon [Common optimal control init for the direct method](https://github.com/control-toolbox/OptimalControl.jl/issues/120).
- [ ] @ocots @PierreMartinon [Replace the print callback of ipopt](https://github.com/control-toolbox/OptimalControl.jl/issues/55).
- [ ] @ocots @BaptisteCbl @joseph-gergaud Benchmark of non linear solvers for shooting
- [ ] @ocots @BaptisteCbl @joseph-gergaud @PierreMartinon Benchmark of optimization solver like `Percival` for the direct method.
- [ ] @joseph-gergaud @PierreMartinon [Lagrangian vs Mayer cost](https://github.com/control-toolbox/CTDirect.jl/issues/35).
- [ ] @joseph-gergaud @PierreMartinon [Non uniform grid](https://github.com/control-toolbox/OptimalControl.jl/issues/133)

**Comment by PierreMartinon** [comment]
Number 2 (constraints) is almost done ? I am still trying to have at least one manual test (goddard) that checks the sign and complementarity conditions.

**Reply by ocots** [reply]
I forgot to mention to improve the plot function about constraints.

**Comment by ocots** [comment]
# Petit point (en fr pour aller plus vite) de ce que j'ai fait et compte faire

## CTBase

- je fais une refonte des fonctions dans `CTBase` pour : i) laisser au user la responsabilité de la cohérence, ii) avoir un code qui ne distingue plus le cas scalaire du cas vectoriel, iii) être un peu plus efficace pour la gestion des cas autonome/non-autonome. Cependant, pour ne pas tout changer d'un coup, je laisse la possibilité en interne de faire un appel spécifique aux fonctions qui se veut être générique au sens où si on fournit le temps (avec un type interne `_Time`) alors le comportment de la fonction est toujours le même, c-a-d on doit fournir des vecteurs et cela retourne des vecteurs même si l'entrée ou la sortie est de dimension 1.
- je mettrai ensuite à jour le `Model`
- j'ajouterai des fonctions `Lie`, `Poisson`, `Lift` qui prendront en argument des `VectorField`... afin de pouvoir gérer des cas d'usage qui pour le moment ne sont pas gérés et que l'on retrouve dans la liste des problèmes de `CTProblems`, par exemple dans le cas d'une contrainte d'état d'ordre 2.
- il faut que je revoie l'affichage du modèle, la gestion des dimensions, des labels...
- il faut être un peu plus défensif avec des messages clairs pour le user s'il a oublié quelque chose ou a mal utilisé une fonction.
- je dois revoir le plot avec la gestion des contraintes, et le `plot!`
- puis gestion de l'init et des callbacks dans `CTBase`.

## CTFlows

- utiliser les nouvelles fonctions en laissant la responsabilité de la cohérence au user 
- ode sur des matrices : Ricatti
- flot d'ocp avec un contrôle, une contrainte et un multiplicateur : gérer cas mixte et état pur correctement
- faire la doc et ajouter des exemples
- voir les gestion du plot : création d'une solution d'un ocp, plot d'un flot hamiltonien...

## CTProblems

- continuer à ajouter des tests unitaires pour vérifier les problèmes ajoutés par @BaptisteCbl 
- finaliser la doc

## CTDirect

- utiliser `_Time` dans un premier temps
- init
- callback

## OptimalControl

- enlever DirectShooting
- ajouter exemples et doc : abstrait + fonctionnel
- ajouter tutos : plot, solveurs, définition d'un modèle...

**Reply by jbcaillau** [reply]
@ocots Good news to simplify the scalar / vectorial case. I am actually a bit worried that some macros for the remaining (relevant) case of autonomous / nonautonomous do an `@eval` (which actually end up with an `eval`, running into well described difficulties for [generated functions](https://github.com/orgs/control-toolbox/discussions/22#discussioncomment-5622604)). Everything seems to work, though (see, *e.g.*, excerpt from `test_onepass.jl` below.
![IMG_2041](https://user-images.githubusercontent.com/62183989/233498499-b6c554fb-aa9d-479c-bd1a-dc073ec04073.jpeg)

**Comment by BaptisteCbl** [comment]
# Petit point

## CTProblems

- j'ai terminé d'ajouter des problèmes dans CTProblems 

## NLSolversBench

- effectuer les benchmarks pour les différents algos de NLsolve, MINPACK, SCIML, ... sur des problèmes de CTProblems afin de déterminer quel(s) algo(s) sont les plus adaptés

## CTBase

- finir de faire le changement de UnitRange à OrdinalRange pour spécifier les contraintes et faire la PR

**Reply by joseph-gergaud** [reply]
Sur CTDirect : 
- Discrétisation par les méthodes de Runge-Kutta
- Docstring des fonctions

**Comment by jbcaillau** [comment]
On abstract formulation, see
- [this issue](https://github.com/control-toolbox/CTBase.jl/issues/9) for ongoing dev with @j-l-s 
- #22 for related questions with `julia`

**Comment by jbcaillau** [comment]
@ocots @PierreMartinon @joseph-gergaud For today (= Friday, April 21, 15:30) catch-up, I would like to discuss with / update you on the following points:
- `CTBase` function refactoring
- `CTProblems` summary
- solver benchmark
- [abstract formulation](https://github.com/control-toolbox/CTBase.jl/issues/9#issuecomment-1515516824)
- [case of variables](https://github.com/control-toolbox/CTBase.jl/issues/9#issuecomment-1515516824)
- `control-toolbox` web
- update general structure including gallery, bocop, hampath, nutopy [^1]
- [Prix Science Ouverte du Logiciel de Recherche (deadline 15/5)](https://www.ouvrirlascience.fr/les-candidatures-pour-le-prix-science-ouverte-du-logiciel-libre-de-la-recherche-2023-sont-ouvertes)
- JuliaCon

[Zoom link](https://univ-cotedazur.zoom.us/my/caillau)

[^1]: move some repos from gitlab to github
